### PR TITLE
Google login

### PR DESCRIPTION
**Pull request by ckhmer1**
Hi,
this request is to integrate the Google login in the project.
It is allowed to login only to one user for now. In order to allow more users to login, we need to update the access and refresh token part.

I hope you want to merge it.
Thanks
Claudio

### CODE REVIEW

File: lib/Auth.js [hunk 1]
```diff
@@ -152,19 +192,38 @@ class Auth {
...
+        let toDel = [];
+        for (let authCode of this._authCode.keys()) {
+            const authCodeInfo = this._authCode.get(authCode);
+            const expirationDate = new Date(authCodeInfo.expiresAt);
+            if (expirationDate < now) {
+                toDel.push(authCode);
+            }
+        }
+        for (let authCode of toDel) {
+            this._authCod.delete(authCode);
```
Comment (Caprico85) [inline lib/Auth.js:221]: `this._authCod` or `this._authCode`?

Comment (ckhmer1) [reply lib/Auth.js]: Yes, sorry, this is a typo.
I should fix it.
Claudio

File: lib/Auth.js [hunk 2]
```diff
@@ -118,19 +121,56 @@ class Auth {
...
+    //
+    getGoogleClientEmails() {
+        return this._emails;
+    }
+    //
+    //
+    //
+    isGoogleClientEmailValid(email) {
+        if (this._useGoogleClientAuth) {
+            const emails = this._emails.split(';');
```
Comment (Caprico85) [inline lib/Auth.js:159]: The Readme says, it's a comma separated list. But here you split by semicolons.

Comment (ckhmer1) [reply lib/Auth.js:136]: Sorry, my mistake. It was my first idea. I've fixed it.

**Comment by ckhmer1** [comment]
Hi,
I've added the multi-user token management.
I hope you merge it.
Best regards
Claudio

File: README.md [hunk 3]
```diff
@@ -425,7 +437,13 @@ If `topic` is something else then `payload` must be an object and tells both the
 
 **Local Authentication**
 
-  `Username` and `Password`: A username and password used when you link Google SmartHome to this node.
+  `Use Google Login`: If enabled, use the Google login authentication.
+
+  `Login Client ID`: If Google Login is enabled, The client id you gained from the *Google Sign-In* integration..
+
+  `Authorized emails`: If Google Login is enabled, The email, comma-separated, authorized to log in.
```
Comment (Caprico85) [inline README.md:444]: Still says "comma-separated" here.

File: lib/Auth.js (hunk 2 above)
Comment (ckhmer1) [reply lib/Auth.js:136]: Sorry again, I've changed the locales but I forgot the README.

File: README.md (hunk 3 above)
Comment (ckhmer1) [reply README.md]: Sorry again, I've changed the locales but I forgot the README.

File: README.md [hunk 4]
```diff
@@ -106,6 +106,18 @@ The HomeGraph API is used to report the state of your devices to Google and to r
...
+
+The Google Sign-In feature allows to login using the Google credential. See [Integrating Google Sign-In into your web app](https://developers.google.com/identity/sign-in/web/sign-in) for more datails.
+
+
+1. Navigate to the [Google Cloud Console API & Services page](https://console.cloud.google.com/apis/credentials)
+2. Click Create credentials > OAuth client ID.
+3. Select the Web application application type.
+4. Name your OAuth 2.0 client and click Create
+
+After configuration is complete, take note of the client ID that was created. You will need the client ID to complete the configuration.
```
Comment (Caprico85) [inline README.md:119]: I had to add my hostname (https://example.com:3001) as authorized JavaScript origin on the Google Credentials page. Maybe you can add this to the Readme.

Comment (ckhmer1) [reply README.md:119]: Done, thanks.

File: lib/Auth.js [hunk 5]
```diff
@@ -249,27 +298,46 @@ class Auth {
...
     isValidRefreshToken(refreshToken) {
-        if (this._auth.refreshToken === refreshToken) {
-            return true;
-        } else {
-            this.debug('Auth:isValidRefreshToken(): refreshToken = ' + JSON.stringify(refreshToken));
-            this.debug('Auth:isValidRefreshToken(): this._auth = ' + JSON.stringify(this._auth));
-            this.debug('Auth:isValidRefreshToken(): invalid refreshToken');
+        const refreshTokenInfo = this._auth.refreshTokens[refreshToken];
+        if (typeof refreshTokenInfo === 'undefined') {
+            this.debug('Auth:isValidRefreshToken(): refreshToken not found ' + JSON.stringify(accessToken));
```
Comment (Caprico85) [inline lib/Auth.js:329]: `JSON.stringify(accessToken))` ... accessToken is not defined here.

Comment (ckhmer1) [reply lib/Auth.js]: I'm sorry about that. I've fixed it.

**Comment by Caprico85** [comment]
Thank you.

I love the Google Login. Finally no more regularly entering my extra secure (and extra long) password.

**Comment by ckhmer1** [comment]
Thanks, I love it too.

**Pull request merged**